### DISCUSSION TITLE
Custom Services and Characteristics

### DISCUSSION THREAD

**Discussion by kfreitag1** [HomeSpan/HomeSpan#78] [Q&A] [answered]
Hey, awesome project! Was just wondering if it is currently possible to create your own services and characteristics using the API?

**Comment by HomeSpan** [comment, marked as answer]
The HomeSpan API itself does not have any specific methods for creating a new HAP Service or HAP Characteristic, but the code where these are defined can easily be modified to add new Services and Characteristics.  That said, Apple's Home App and Siri only recognize Apple-defined Services and Characteristics, so unless you are adding something to HomeSpan that is defined by Apple but not included in HomeSpan, any new Services or Characteristics will be ignored by the Home App and Siri.  There is one set of Apple Services and Characteristics that could be supported by HomeSpan but are not included: Apple's control for Televisions.  Unfortunately, Apple has not released the specs for these so I can't add them to HomeSpan, though others have done so using information gleaned from the web.

I believe the reason HAP allows you to create new Services or Characteristics even though they will be ignored by the Home App and Siri is that the Home App is not the only HomeKit App.  There are others, and these could read and process non-Apple Services and Characteristics.  This also allows manufacturers to create a custom app to control a specific device that is HomeKit compatible but offers more control features in their custom app than in the Home App.

**Reply by kfreitag1** [reply]
Ah yes that makes sense, thank you for the detailed response!
